File: pkgs/dns/dns_resolve.js
// DNS
// ===

// A numeric host is the shared io_addr rule. A name is getaddrinfo through
// bun:ffi (Bun has no dns.lookupSync, and an async lookup cannot run while
// the kit is inside libc poll). The C lane uses the same call on a helper
// thread via io_work. This lane blocks for the whole lookup.
function resolve(host) {
  if (host.length === 0 || host.includes("\0")) {
    return io_fail(22);
  }
  if (io_addr(host, 0) !== null) {
    return io_done(host);
  }
  // Hex / short forms: inet_aton accepts what io_addr refuses.
  if (dns_inet_aton(host)) {
    return io_fail(22);
  }
  // Digits and dots only: refuse before lookup (parity with C).
  if (/^[0-9.]+$/.test(host)) {
    return io_fail(22);
  }
  try {
    return io_done(dns_getaddrinfo_v4(host));
  } catch (e) {
    if (typeof e !== "number") {
      throw e;
    }
    return io_fail(e >>> 0);
  }
}

function dns_lib() {
  if (globalThis.BEND_DNS === undefined) {
    const ffi = require("bun:ffi");
    const mac = process.platform === "darwin";
    const lib = ffi.dlopen(mac ? "libSystem.dylib" : "libc.so.6", {
      getaddrinfo: { args: ["ptr", "ptr", "ptr", "ptr"], returns: "i32" },
      freeaddrinfo: { args: ["ptr"], returns: "void" },
      inet_ntop: { args: ["i32", "ptr", "ptr", "u32"], returns: "cstring" },
      inet_aton: { args: ["ptr", "ptr"], returns: "i32" },
    });
    globalThis.BEND_DNS = { ...lib.symbols, ptr: ffi.ptr, read: ffi.read, mac };
  }
  return globalThis.BEND_DNS;
}

function dns_cstr(s) {
  return dns_lib().ptr(new TextEncoder().encode(s + "\0"));
}

function dns_inet_aton(host) {
  const lib = dns_lib();
  const out = new Uint8Array(4);
  return lib.inet_aton(dns_cstr(host), lib.ptr(out)) !== 0;
}

function dns_getaddrinfo_v4(host) {
  const lib = dns_lib();
  const mac = lib.mac;
  // struct addrinfo: Linux puts ai_addr before ai_canonname; Darwin the reverse.
  const hints = new Uint8Array(48);
  const hv = new DataView(hints.buffer);
  hv.setInt32(4, 2, true); // ai_family = AF_INET
  hv.setInt32(8, 1, true); // ai_socktype = SOCK_STREAM
  const out = new BigUint64Array(1);
  const rc = lib.getaddrinfo(dns_cstr(host), null, lib.ptr(hints), lib.ptr(out));
  if (rc !== 0) {
    // EAI_NONAME is 8 on Darwin, -2 on glibc; EAI_AGAIN is 2 / -3.
    const code = (mac ? rc === 8 : rc === -2) ? 2
      : (mac ? rc === 2 : rc === -3) ? (mac ? 35 : 11)
      : 22;
    throw code;
  }
  const res = Number(out[0]);
  try {
    const ai_addr_off = mac ? 32 : 24;
    const sockaddr = Number(lib.read.ptr(res + ai_addr_off));
    const sin_addr = sockaddr + 4;
    const buf = new Uint8Array(16);
    const text = lib.inet_ntop(2, sin_addr, lib.ptr(buf), 16);
    if (!text) {
      throw 22;
    }
    return String(text);
  } finally {
    lib.freeaddrinfo(res);
  }
}

File: pkgs/dns/dns_resolve.c
// DNS
// ===

#include <netdb.h>

// getaddrinfo's EAI_* codes are not errno. Map the ones a program will see
// onto the same codes the rest of the kit uses (ENOENT / EAGAIN / EINVAL).
static u32 dns_eai(int rc) {
  if (rc == EAI_SYSTEM) {
    return errno != 0 ? (u32)errno : EINVAL;
  }
  if (rc == EAI_AGAIN) {
    return EAGAIN;
  }
#if defined(EAI_NONAME)
  if (rc == EAI_NONAME) {
    return ENOENT;
  }
#endif
#if defined(EAI_NODATA)
  if (rc == EAI_NODATA) {
    return ENOENT;
  }
#endif
  return EINVAL;
}

// Digits and dots only: a failed io_sys_addr must not fall through to
// getaddrinfo (macOS accepts leading-zero quads that the kit rejects).
static int dns_numeric_shape(const char* host) {
  if (host[0] == 0) {
    return 0;
  }
  for (const char* p = host; *p != 0; p += 1) {
    if (!((*p >= '0' && *p <= '9') || *p == '.')) {
      return 0;
    }
  }
  return 1;
}

static void resolve_call(IoWork* w) {
  struct addrinfo hints, *res = NULL;
  memset(&hints, 0, sizeof hints);
  hints.ai_family = AF_INET;
  hints.ai_socktype = SOCK_STREAM;
  int rc = getaddrinfo(w->data, NULL, &hints, &res);
  if (rc != 0) {
    w->code = dns_eai(rc);
    return;
  }
  struct sockaddr_in* sin = (struct sockaddr_in*)res->ai_addr;
  if (inet_ntop(AF_INET, &sin->sin_addr, w->text, INET_ADDRSTRLEN) == NULL) {
    w->code = errno != 0 ? (u32)errno : EINVAL;
    freeaddrinfo(res);
    return;
  }
  freeaddrinfo(res);
  w->code = 0;
}

static Term resolve_pack(Env e, IoWork* w) {
  free(w->data);
  if (w->code != 0) {
    free(w->text);
    return io_fail(e, w->code, NULL);
  }
  Term r = io_done(e, io_str(e, w->text, strlen(w->text)));
  free(w->text);
  return r;
}

Term resolve_run(Env e, Term* f, IoWork* w) {
  struct sockaddr_in at;
  struct in_addr soft;
  char out[INET_ADDRSTRLEN];
  w->data = io_cstr(e, f[0], &w->size);
  if (io_nul(w->data, w->size) || w->size == 0) {
    free(w->data);
    return io_fail(e, EINVAL, NULL);
  }
  // Fast path: the same canonical-quad rule TCP.connect / UDP.send_to use.
  if (io_sys_addr(w->data, 0, &at) == 0) {
    if (inet_ntop(AF_INET, &at.sin_addr, out, sizeof out) == NULL) {
      free(w->data);
      return io_fail(e, errno != 0 ? (u32)errno : EINVAL, NULL);
    }
    free(w->data);
    return io_done(e, io_str(e, out, strlen(out)));
  }
  // Hex / short forms (0x7f000001, 127.1): getaddrinfo's inet_aton front
  // end would accept them; TCP.connect refuses. Keep the shape guard too —
  // inet_aton rejects 1.2.3.4.5 / 999.1.1.1 that must stay EINVAL.
  if (inet_aton(w->data, &soft)) {
    free(w->data);
    return io_fail(e, EINVAL, NULL);
  }
  if (dns_numeric_shape(w->data)) {
    free(w->data);
    return io_fail(e, EINVAL, NULL);
  }
  w->text = io_mem(malloc(INET_ADDRSTRLEN));
  return io_work(w, resolve_call, resolve_pack);
}

static void __attribute__((constructor)) resolve_use(void) {
  io_eff(CID_RESOLVE, resolve_run, 0);
}
